crates/oxc_semantic/src/checker/javascript.rs

```diff
@@ -108,7 +108,14 @@ fn invalid_let_declaration(x0: &str, span1: Span) -> OxcDiagnostic {
 pub fn check_binding_identifier(ident: &BindingIdentifier, ctx: &SemanticBuilder<'_>) {
     let strict_mode = ctx.strict_mode();
     // It is a Diagnostic if the StringValue of a BindingIdentifier is "eval" or "arguments" within strict mode code.
-    if strict_mode && matches!(ident.name.as_str(), "eval" | "arguments") {
+    if strict_mode
+        && matches!(ident.name.as_str(), "eval" | "arguments")
+        && ctx
+            .nodes
+            .parent_kind(ctx.current_node_id)
+            .as_function()
+            .is_none_or(|function| !matches!(function.r#type, FunctionType::TSDeclareFunction))
+    {
```

Comment:
I'd assumed `declare function f({eval});` would also be legal. But it seems not: [TS Playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIAoQA3KCALnlWQFsAjEGAGnlgHNaRUMBnS6+owCU-WgxgBuAFChIsBCnTY88NgQDeJMi3aduPAL6VNpClTGMdMDjS69RgmAZHnH02dDhI0mXPgAWBADaWhBWNnY8ALqUQQLiLPGMUS5JklJAA)

It might be worth making a comment to that effect to explain why only checking the parent and not searching further up ancestry.

Reply:
Oh wait! This PR relates to `declare function eval();` not `declare function f(eval);`. Presumably we handle that one too.